src/dbnode/storage/namespace_bootstrap_data_accumulator.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import (
+	"github.com/m3db/m3/src/dbnode/storage/bootstrap"
+	"github.com/m3db/m3/src/dbnode/storage/series/lookup"
+	"github.com/m3db/m3/src/x/ident"
+	"github.com/m3db/m3/src/x/instrument"
+)
+
+type namespaceDataAccumulator struct {
+	namespace    databaseNamespace
+	needsRelease []lookup.OnReleaseReadWriteRef
+}
+
+func newDatabaseNamespaceDataAccumulator(
+	namespace databaseNamespace,
+) bootstrap.NamespaceDataAccumulator {
+	return &namespaceDataAccumulator{
+		namespace: namespace,
+	}
+}
+
+func (a *namespaceDataAccumulator) CheckoutSeries(
+	id ident.ID,
+	tags ident.TagIterator,
+) (bootstrap.CheckoutSeriesResult, error) {
+	ref, err := a.namespace.SeriesReadWriteRef(id, tags)
+	if err != nil {
+		return bootstrap.CheckoutSeriesResult{}, err
+	}
+
+	a.needsRelease = append(a.needsRelease, ref.ReleaseReadWriteRef)
+	return bootstrap.CheckoutSeriesResult{
+		Series:      ref.Series,
+		UniqueIndex: ref.UniqueIndex,
+	}, nil
+}
+
+func (a *namespaceDataAccumulator) Release() {
+	// Release all refs.
+	for _, elem := range a.needsRelease {
+		elem.OnReleaseReadWriteRef()
+	}
+
+	// Memset optimization for reset.
+	for i := range a.needsRelease {
+		a.needsRelease[i] = nil
+	}
+	a.needsRelease = a.needsRelease[:0]
+}
+
+func (a *namespaceDataAccumulator) Close() error {
```

Comment:
TODO: Move release into Close() and make sure defer Close() is called from where `newDatabaseNamespaceDataAccumulator` is called (i.e. accumulator is created) since right now it is not.

Reply:
Just add locks, much faster than the lookup happening `SeriesReadWriteRef(....)`

Reply:
TODO: Name methods and add documentation to make it clear only single threaded use allowed.

Reply:
Yup, I'm adding a lock at that place.